core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -408,11 +450,15 @@ private TableMetadata applyChangesToMapping(TableMetadata metadata) {
   private static Schema applyChanges(Schema schema, List<Integer> deletes,
                                      Map<Integer, Types.NestedField> updates,
                                      Multimap<Integer, Types.NestedField> adds,
-                                     Multimap<Integer, Move> moves) {
+                                     Multimap<Integer, Move> moves,
+                                     Set<String> rowIdentifierNames) {
     Types.StructType struct = TypeUtil
         .visit(schema, new ApplyChanges(deletes, updates, adds, moves))
         .asNestedType().asStructType();
-    return new Schema(struct.fields());
+    Set<Integer> refreshedRowIdentifiers = rowIdentifierNames.stream()
+        .map(name -> struct.field(name).fieldId())
```

Comment:
If we delete column `b` by using the `SchemaUpdate#deleteColumn` API,  then the `struct` variable which has been applied by those changes  won't include the field that has the colum name `b`, right ?   If sure, then how could we locate the field id for that given column name `b` in this sentence ?

Reply:
Yes, that is why I decide to throw exception when `deleteColumn` is called and the column is a row identifier. There are 2 possible situations:

1. the user actually wants to delete the column, then he/she should call:

```
update.deleteRowIdentifier(columnName).deleteColumn(columnName);
```

2. the user does not know the column is used as a row identifier, and in that case just calling `deleteColumn` would result in exception.

Reply:
Shoud we add the unit tests for above cases ?  I mean we are deleting the rowIdentifierField and columns in the same txn .

Reply:
Yes, we should definitely have a test for this case.